### PR TITLE
[stdlib] Use merge sort for sortWith

### PR DESCRIPTION
**Pull request by ilmat192**

### CODE REVIEW

File: runtime/src/main/kotlin/kotlin/util/Sort.kt [hunk 1]
```diff
@@ -339,7 +363,7 @@ private fun quickSort(
 internal fun <T> sortArrayWith(
         array: Array<out T>, fromIndex: Int = 0, toIndex: Int = array.size, comparator: Comparator<T>) {
     @Suppress("UNCHECKED_CAST")
-    quickSort(array as Array<T>, fromIndex, toIndex - 1, comparator)
+    mergeSort(array as Array<T>, fromIndex, toIndex - 1, comparator)
```
Comment (sbogolepov) [inline runtime/src/main/kotlin/kotlin/util/Sort.kt]: Maybe change the implementation of `sortArrayComparable` as well?

Comment (ilmat192) [reply runtime/src/main/kotlin/kotlin/util/Sort.kt]: Fixed

File: backend.native/tests/runtime/collections/SortWith.kt [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+
+// Assert that the array is sorted in terms of a comparator only for correct/partially correct cases
+val comparators = listOf<ComparatorInfo>(
+    ComparatorInfo("Correct increasing", correctIncreasing ,  true),
+    ComparatorInfo("Correct decreasing", correctDecreasing ,  true),
+    ComparatorInfo("Incorrect increasing", Comparator { a ,b -> if (a > b) 1 else -1 },  false),
+    ComparatorInfo("Incorrect decreasing", Comparator { a ,b -> if (a < b) 1 else -1 },  false),
+    ComparatorInfo("Always 1",  Comparator { a ,b ->  1  },  false),
+    ComparatorInfo("Always -1", Comparator { a ,b ->  -1 },  false),
+    ComparatorInfo("Alwasy 0",  Comparator { a ,b ->  0  },  false)
```
Comment (olonho) [inline backend.native/tests/runtime/collections/SortWith.kt]: Always

File: runtime/src/main/kotlin/kotlin/util/Sort.kt [hunk 3]
```diff
@@ -20,67 +20,114 @@ import kotlin.comparisons.*
...
-        if (i <= j) {
-            val tmp = array[i]
-            array[i] = array[j]
-            array[j] = tmp
-            i++
-            j--
+    val buffer = arrayOfNulls<Any?>(array.size) as Array<T>
+    val result = mergeSort(array, buffer, start, endInclusive)
+    if (result !== array) {
+        result.forEachIndexed { i, v-> array[i] = v }
```
Comment (olonho) [inline runtime/src/main/kotlin/kotlin/util/Sort.kt]: i, v -> array[i] = v
   ^^^ space

File: runtime/src/main/kotlin/kotlin/util/Sort.kt [hunk 4]
```diff
@@ -20,67 +20,114 @@ import kotlin.comparisons.*
 
 // TODO: Implement sort for primitives with a custom comparator.
 
-// Yes, rather naive qsort for now.
 // Array<T>     =============================================================================
-private fun <T> partition(array: Array<T>, left: Int, right: Int): Int {
-    var i = left
-    var j = right
+// We use merge becuase the quick sort may cause segfaults if
```
Comment (olonho) [inline runtime/src/main/kotlin/kotlin/util/Sort.kt]: because

File: runtime/src/main/kotlin/kotlin/util/Sort.kt [hunk 5]
```diff
@@ -20,67 +20,114 @@ import kotlin.comparisons.*
...
+    return target
 }
 
-private fun <T> quickSort(array: Array<T>, left: Int, right: Int) {
-    val index = partition(array, left, right)
-    if (left < index - 1)
-        quickSort(array, left, index - 1)
-    if (index < right)
-        quickSort(array, index, right)
+// Sort with comparator
```
Comment (olonho) [inline runtime/src/main/kotlin/kotlin/util/Sort.kt]: Do at the end.

File: runtime/src/main/kotlin/kotlin/util/Sort.kt [hunk 6]
```diff
@@ -20,67 +20,114 @@ import kotlin.comparisons.*
...
+        return array
+    }
+
+    val median = (start + end) / 2
+    val left = mergeSort(array, buffer, start, median, comparator)
+    val right = mergeSort(array, buffer, median + 1, end, comparator)
+
+    val target = if (left === buffer) array else buffer
+
+    // Merge
```
Comment (olonho) [inline runtime/src/main/kotlin/kotlin/util/Sort.kt]: Dot at the end.

**Pull request merged**